app/uk/gov/hmrc/emailverification/controllers/EmailPasscodeController.scala

```diff
@@ -0,0 +1,118 @@
+import uk.gov.hmrc.emailverification.models._
+import uk.gov.hmrc.emailverification.repositories.{PasscodeMongoRepository, VerifiedEmailMongoRepository}
+import uk.gov.hmrc.http.{HeaderCarrier, UpstreamErrorResponse}
+import uk.gov.hmrc.play.audit.AuditExtensions._
+import uk.gov.hmrc.play.audit.http.connector.AuditConnector
+import uk.gov.hmrc.play.audit.model.ExtendedDataEvent
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Random
+
+
+class EmailPasscodeController @Inject()(emailConnector: EmailConnector,
+                                        passcodeRepo:PasscodeMongoRepository,
+                                        verifiedEmailRepo: VerifiedEmailMongoRepository,
+                                        analyticsConnector: PlatformAnalyticsConnector,
+                                        auditConnector: AuditConnector,
+                                        controllerComponents: ControllerComponents
+                                            )(implicit ec:ExecutionContext, appConfig: AppConfig) extends BaseControllerWithJsonErrorHandling(controllerComponents) {
+
+  def newPasscode(): String =  {
+    val codeSize = 6
+    Random.alphanumeric
+      .filterNot(_.isDigit)
+      .filterNot(_.isLower)
+      .filterNot(Set('A', 'E', 'I', 'O', 'U'))
+      .take(codeSize).mkString
+  }
+
+  private def sendEmailAndStorePasscode(request: PasscodeRequest)(implicit hc: HeaderCarrier) = {
+    val passcode = newPasscode()
+    val paramsWithPasscode = request.templateParameters.getOrElse(Map.empty) +
+      ("passcode" -> passcode)
+
+    for {
+      _ <- emailConnector.sendEmail(request.email, request.templateId, paramsWithPasscode)
+      _ <- passcodeRepo.upsert(request.sessionId, passcode, request.email, request.linkExpiryDuration)
+    } yield Created
+  }
+
+  def requestPasscode(): Action[JsValue] = Action.async(parse.json) {
+    implicit httpRequest =>
+      withJsonBody[PasscodeRequest] { request =>
+        verifiedEmailRepo.isVerified(request.email) flatMap {
+          case true => Future.successful(Conflict(Json.toJson(ErrorResponse("EMAIL_VERIFIED_ALREADY", "Email has already been verified"))))
+          case false =>
+            analyticsConnector.sendEvents(GaEvents.passcodeRequested)
+            sendEmailAndStorePasscode(request)
+        } recover {
+          case ex @ UpstreamErrorResponse(_, 400, _, _) =>
+            val event = ExtendedDataEvent(
+              auditSource =  "email-verification",
+              auditType = "AIV-60",
+              tags = hc.toAuditTags("requestPasscode", httpRequest.path),
+              detail = Json.obj(
+                "email-address" -> request.email,
+                "email-address-hex" -> toByteString(request.email)
+              )
+            )
+            auditConnector.sendExtendedEvent(event)
+            Logger.error("email-verification had a problem reading from repo", ex)
```

Comment:
surely the repository isn't throwing UpstreamErrorResponse on failures?

Reply:
wooaahh good spot, copied from earlier implementation so that's wrong as well